segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/WriteOperation.java

```diff
@@ -417,23 +435,49 @@ private CompletableFuture<Void> writeToChunk(MetadataTransaction txn,
                     .thenApplyAsync(h -> bytesCount, chunkedSegmentStorage.getExecutor());
         }
         return retValue
+                .thenComposeAsync(bytesWritten -> {
+                    // Validate.
+                    Preconditions.checkState(bytesWritten == bytesCount,
+                            "bytesWritten (%s) must equal bytesCount(%s). Segment=%s Chunk=%s offsetToWriteAt=%s",
+                            bytesWritten, bytesCount, segmentMetadata, chunkWrittenMetadata, offsetToWriteAt);
+
+                    CompletableFuture<Void> validation = CompletableFuture.completedFuture(null);
+                    if (chunkedSegmentStorage.getConfig().isSelfCheckForMetadataEnabled()) {
+                        validation = validation.thenComposeAsync(v ->
+                                chunkedSegmentStorage.getChunkStorage().getInfo(chunkHandle.getChunkName())
+                                    .thenAcceptAsync( chunkInfo -> {
+                                        Preconditions.checkState(chunkInfo.getLength() == (chunkWrittenMetadata.getLength() + bytesWritten),
```

Comment:
In the case this happens, wouldn't it be more accurate to throw a `DataCorruptionException`? It would be easier for us to detect, at least while using this feature for debug purposes. My understanding is that once this happens, this write will be discarded, garbage collected and it would be re-attempted, right?

Reply:
DataCorruptionException is defined in io.pravega.segmentstore.server. Storage can not depend on that project.